File: NeetCodeIO/131PalindromePartitioning/PalindromePartitioning.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        # Backtracking -> Time:O(2^n), Space:O(n)
        ans = []
        part = []
        def dfs(i):
            if i >= len(s):
                ans.append(part.copy())
                return
            for j in range(i, len(s)):
                if s[i:j+1] == s[i:j+1][::-1]:
                    part.append(s[i:j+1])
                    dfs(j+1)
                    part.pop()
        dfs(0)
        return ans

File: NeetCodeIO/131PalindromePartitioning/PalindromePartitioning.cpp
class Solution {
public:

    vector<vector<string>> ans;
    vector<string> part;

    bool isPalin(string s) {
        for (int i = 0; i < s.size()/2; i++) {
            if (s[i] != s[s.size()-1-i]) {
                return false;
            }
        }
        return true;
    }

    void dfs(int i, string s) {
        if (i >= s.size()) {
            ans.push_back(part);
            return;
        }
        for (int j = i; j < s.size(); j++) {
            if (isPalin(s.substr(i, j-i+1))) {
                part.push_back(s.substr(i, j-i+1));
                dfs(j+1, s);
                part.pop_back();
            }
        }
    }

    vector<vector<string>> partition(string s) {
        // Backtracking -> Time:O(2^n), Space:O(n)
        dfs(0, s);
        return ans;
    }
};
